### ISSUE TITLE
MPS and cpu method results are drastically different

### ISSUE THREAD

**Issue by dcaustin33** [pytorch/pytorch#131285]
### 🐛 Describe the bug

Using MPS with the grounding dino implementation (https://github.com/IDEA-Research/GroundingDINO) leads to extremely different results (MPS first pic, cpu second pic). The image can be found at https://media.roboflow.com/notebooks/examples/dog.jpeg
And the code I am using is as follows, with no source code adjustments in grounding dino. All I am changing is the device.

```from groundingdino.util.inference import load_model, load_image, predict, annotate
import cv2
import torch

model = load_model("groundingdino/config/GroundingDINO_SwinT_OGC.py", "/Users/derek/Desktop/GroundingDINO/groundingdino_swint_ogc.pth")
IMAGE_PATH = "/Users/derek/Desktop/GroundingDINO/dog.jpeg"
TEXT_PROMPT = "dog"
# TEXT_PROMPT = "tennis player"
BOX_TRESHOLD = 0.15
TEXT_TRESHOLD = 0.25

image_source, image = load_image(IMAGE_PATH)

boxes, logits, phrases = predict(
    model=model,
    image=image,
    caption=TEXT_PROMPT,
    box_threshold=BOX_TRESHOLD,
    text_threshold=TEXT_TRESHOLD,
    device=torch.device("mps")
)

annotated_frame = annotate(
    image_source=image_source, boxes=boxes, logits=logits, phrases=phrases)
cv2.imwrite("annotated_image.jpg", annotated_frame)
```

I am using torch version 2.3.1

<img width="584" alt="Screenshot 2024-07-20 at 4 19 49 PM" src="https://github.com/user-attachments/assets/7f1c5e52-8d9f-4887-bf85-af4e018709bd">
<img width="588" alt="Screenshot 2024-07-20 at 4 20 05 PM" src="https://github.com/user-attachments/assets/ba6e8c22-c420-4f8d-915e-9fa4f7767edb">

### Versions

`Collecting environment information...
PyTorch version: 2.3.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 14.4 (arm64)
GCC version: Could not collect
Clang version: 15.0.0 (clang-1500.3.9.4)
CMake version: Could not collect
Libc version: N/A

Python version: 3.10.14 (main, Jul 20 2024, 13:03:49) [Clang 15.0.0 (clang-1500.3.9.4)] (64-bit runtime)
Python platform: macOS-14.4-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M3 Max

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] torch==2.3.1
[pip3] torchvision==0.18.1
[conda] numpy                     1.26.4          py312h7f4fdc5_0  
[conda] numpy-base                1.26.4          py312he047099_0  
[conda] numpydoc                  1.7.0           py312hca03da5_0  `

cc @kulinseth @albanD @malfet @DenisVieriu97 @jhavukainen

**Comment by dcaustin33** [comment]
I have traced it back to a masked_fill operation that should not change any values in the implementation but it does in this case. Commenting out that line for a single inference example gives the right response but with multiple images I believe we may need this operation so I think its a big issue

https://github.com/IDEA-Research/GroundingDINO/blob/e27a646ca0ea29c7883187c6f7303cc975366679/groundingdino/models/GroundingDINO/fuse_modules.py#L209C13-L209C27

**Comment by qqaatw** [comment]
Can you print out the inputs, i.e. `attention_mask_l`, to `masked_fill_` that causes incorrect results?

**Comment by dcaustin33** [comment]
Yes, as you can see mask is all False and attn_weights_l is a rather large tensor. Unsure how I can upload the pt files

<img width="724" alt="Screenshot 2024-07-20 at 10 20 52 PM" src="https://github.com/user-attachments/assets/a7f73a0f-f955-4041-92f4-5eebd997bd0e">

**Comment by qqaatw** [comment]
Thanks for providing the pic, that's very clear. Let me look into it.

**Comment by qqaatw** [comment]
Can you print out `attention_mask_v` as well? (without applying the sum), and also can you print the shapes of them?

**Comment by dcaustin33** [comment]
Yes here they are

<img width="616" alt="Screenshot 2024-07-21 at 11 12 04 AM" src="https://github.com/user-attachments/assets/b2f1b83c-b50a-4c4f-9828-77055289627c">

**Comment by qqaatw** [comment]
Hmm I took a look at your pic, it seems working properly doesn't it? 

`mask_fill_` fills the value when the mask is `True`. Since all the elements in mask are `False`, the values in `attn_weights_l` should be unchanged, which looks correct from your pic. How did you inflate the second dimension of the mask btw? did you also inflate the `attn_weights_l`?

Could you try changing this line and see if the issue persists:

```python
attn_weights_l.masked_fill_(attention_mask_v, float("-inf"))
# to 
attn_weights_l.masked_fill_(attention_mask_v.contiguous(), float("-inf"))
``` 

If so please try uploading the weights file here (just for minimal reproduction), there should be an attach file button on the GH interface.

**Comment by dcaustin33** [comment]
No the above pic was me showing that it works if I inflate the mask dimensions. The weights and the code to reproduce are attached.
[reproduce_error.zip](https://github.com/user-attachments/files/16355895/reproduce_error.zip)



<img width="698" alt="Screenshot 2024-07-23 at 9 12 38 PM" src="https://github.com/user-attachments/assets/4b504133-a8bc-4295-93d4-c10d85fe2df1">

**Comment by qqaatw** [comment]
I couldn't repro it with your script:

```python
import torch

device = "mps"
attn_weights = torch.load("attn_weights.pt").to(device)
attn_mask = torch.load("attn_mask.pt").to(device)

print(attn_weights.shape)
print(attn_mask.shape)
print(attn_mask.sum())

attn_weights.masked_fill_(attn_mask, float("-inf"))

print((attn_weights == float("-inf")).sum())

attn_weights[0][0][0] = float("-inf")

print((attn_weights == float("-inf")).sum())
```

out:
```
torch.Size([8, 5, 19322])
torch.Size([8, 1, 19322])
tensor(0, device='mps:0')
tensor(0, device='mps:0')
tensor(1, device='mps:0')
```

Can you try running it in a regular python runtime? i.e. not using jupyter kernel.

**Comment by dcaustin33** [comment]
Yes I did - if you see from the printouts it is not that any actually go to negative infinity. They instead change value for absolutely no reason. IN the above screenshot look at the second value on the first row for both the first and second printout. They should be identical, but they are not.

**Comment by qqaatw** [comment]
@dcaustin33 ok I got your point, will look into it.

**Comment by qqaatw** [comment]
I got the root cause of this issue. As a temporary fix, you can use the out of place version of `masked_fill` or do a contiguous operation before `masked_fill_`, like so: 

`attn_weights = attn_weights.contiguous().masked_fill_(attn_mask, float("-inf"))`

Relevant fix: https://github.com/pytorch/pytorch/pull/128393, I will need to consult the apple team to discuss the best solution for the in-place case.

**Comment by dcaustin33** [comment]
Awesome thanks so much @qqaatw !
